src/vm/runners/builtin_runner/bitwise.rs

```diff
@@ -111,9 +111,31 @@ impl BitwiseBuiltinRunner {
     }
 
     pub fn get_allocated_memory_units(&self, vm: &VirtualMachine) -> Result<usize, MemoryError> {
-        let value = safe_div_usize(vm.current_step, self.ratio as usize)
-            .map_err(|_| MemoryError::ErrorCalculatingMemoryUnits)?;
-        Ok(self.cells_per_instance as usize * value)
+        match self.ratio {
+            None => {
+                // Dynamic layout has the exact number of instances it needs (up to a power of 2).
+                let instances: usize =
+                    self.get_used_cells(&vm.segments)? / self.cells_per_instance as usize;
+                let components =
+                    (instances / self.instances_per_component as usize).next_power_of_two();
+                Ok(self.cells_per_instance as usize
+                    * self.instances_per_component as usize
+                    * components)
+            }
+            Some(ratio) => {
+                let min_step = (ratio * self.instances_per_component) as usize;
+                if vm.current_step < min_step {
+                    return Err(InsufficientAllocatedCellsError::MinStepNotReached(
+                        min_step,
+                        BITWISE_BUILTIN_NAME,
+                    )
+                    .into());
+                };
+                let value = safe_div_usize(vm.current_step, ratio as usize)
+                    .map_err(|_| MemoryError::ErrorCalculatingMemoryUnits)?;
```

Comment:
We can make `ratio` `NonZeroU32` to avoid this scenario.

Reply:
`safe_div` can still fail if the reminder is not zero

Reply:
Oh, it's a normal division 🤦 
I thought these were `Felt`s. For those only dividing by `0` can fail, everything else can be divided by any other element in the field.